ledger/simulation/simulation_eval_test.go

```diff
@@ -1475,6 +1573,570 @@ int 1`
+								},
+							},
+							{
+								Trace: &simulation.TransactionTrace{},
+							},
+							{
+								Txn: transactions.SignedTxnWithAD{
+									ApplyData: transactions.ApplyData{
+										ApplicationID: 0,
+										EvalDelta: transactions.EvalDelta{
+											Logs: []string{string(uint64ItoB(1 << MaxDepth))},
+											InnerTxns: []transactions.SignedTxnWithAD{
+												{
+													ApplyData: transactions.ApplyData{ApplicationID: futureAppID + 3},
+												},
+												{},
+												{},
+												{},
+												{
+													ApplyData: transactions.ApplyData{
+														EvalDelta: transactions.EvalDelta{
+															Logs: []string{string(uint64ItoB(1 << (MaxDepth - 1)))},
+															InnerTxns: []transactions.SignedTxnWithAD{
+																{
+																	ApplyData: transactions.ApplyData{ApplicationID: futureAppID + 8},
+																},
+																{},
+																{},
+																{},
+																{
+																	ApplyData: transactions.ApplyData{
+																		EvalDelta: transactions.EvalDelta{
+																			Logs: []string{string(uint64ItoB(1 << (MaxDepth - 2)))},
+																		},
+																	},
+																},
+															},
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+								AppBudgetConsumed: 378,
+								Trace: &simulation.TransactionTrace{
+									ApprovalProgramTrace: simulation.ProgramTrace{
+										Trace:          recursiveLongOpcodeTrace,
+										StepToInnerMap: stepToInnerMap,
+									},
+									InnerTraces: []simulation.TransactionTrace{
+										{
+											ApprovalProgramTrace: simulation.ProgramTrace{Trace: creationOpcodeTrace},
+										},
+										{},
+										{
+											ApprovalProgramTrace: simulation.ProgramTrace{Trace: optInTrace},
+										},
+										{
+											ClearStateProgramTrace: simulation.ProgramTrace{Trace: clearStateOpcodeTrace},
+										},
+										{
+											ApprovalProgramTrace: simulation.ProgramTrace{
+												Trace:          recursiveLongOpcodeTrace,
+												StepToInnerMap: stepToInnerMap,
+											},
+											InnerTraces: []simulation.TransactionTrace{
+												{
+													ApprovalProgramTrace: simulation.ProgramTrace{Trace: creationOpcodeTrace},
+												},
+												{},
+												{
+													ApprovalProgramTrace: simulation.ProgramTrace{Trace: optInTrace},
+												},
+												{
+													ClearStateProgramTrace: simulation.ProgramTrace{Trace: clearStateOpcodeTrace},
+												},
+												{
+													ApprovalProgramTrace: simulation.ProgramTrace{Trace: finalDepthTrace},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+						AppBudgetAdded:    4200,
+						AppBudgetConsumed: 385,
+					},
+				},
+				IncludePC: true,
+			},
+		}
+	})
+}
+
+func TestLogicSigPCExposure(t *testing.T) {
```

Comment:
These are good tests. But can you also include tests for a failing app call and a failing logicsig?

Reply:
yes, see [update app failing test and logic sig failing test](https://github.com/algorand/go-algorand/pull/5360/commits/e9025bbb6f1092b5cdac4ca43fc809873d0ff2b7)